tokio-executor/src/park.rs

```diff
@@ -237,37 +238,56 @@ impl Inner {
     fn park(&self, timeout: Option<Duration>) -> Result<(), ParkError> {
         // If currently notified, then we skip sleeping. This is checked outside
         // of the lock to avoid acquiring a mutex if not necessary.
-        match self.state.compare_and_swap(NOTIFY, IDLE, Ordering::SeqCst) {
-            NOTIFY => return Ok(()),
-            IDLE => {},
-            _ => unreachable!(),
+        if self.state.compare_exchange(NOTIFY, IDLE, SeqCst, SeqCst).is_ok() {
+            return Ok(());
+        }
+
+        // If the duration is zero, then there is no need to actually block
+        if let Some(ref dur) = timeout {
```

Comment:
This check is new in `tokio-executor`. @carllerche, any idea why the previous implementation didn't have it? Maybe just an oversight?

Reply:
Oversight... it would be nice to unify the two impls instead of duplicating code.